jdaviz/core/template_mixin.py

```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
+    * ``selected_obj`` (viewer item corresponding to ``selected``)
+
+
+    To use in a plugin:
+
+    * create traitlets with default values
+    * register with all the automatic logic in the plugin's init by passing the string names
+      of the respective traitlets.
+    * use component in plugin template (see below)
+    * refer to properties above based on the interally stored reference to the
+      instantiated object of this component
+
+    Example template (label and hint are optional)::
+
+      <plugin-viewer-select
+        :items="viewer_items"
+        :selected.sync="viewer_selected"
+        label="Viewer"
+        hint="Select viewer."
+      />
+
+    """
+    def __init__(self, plugin, items, selected):
+        super().__init__(plugin,
+                         items=items,
+                         selected=selected)
+
+        self.hub.subscribe(self, ViewerAddedMessage,
+                           handler=lambda _: self._on_viewers_changed())
+        self.hub.subscribe(self, ViewerRemovedMessage,
+                           handler=lambda _: self._on_viewers_changed())
+        self.add_observe(selected, self._selected_changed)
+
+        # initialize viewer_items from original viewers
+        self._on_viewers_changed()
+
+    @property
+    def ids(self):
+        return [item['id'] for item in self.items]
+
+    @property
+    def references(self):
+        return [item['reference'] for item in self.items]
+
+    @property
+    def ref_or_ids(self):
+        return [item['ref_or_id'] for item in self.items]
+
+    @property
+    def selected_item(self):
+        for item in self.items:
+            if item['ref_or_id'] == self.selected:
+                return item
+        # try again but this time allow match to id alone.  Note that _selected_changed
+        # will handle resetting the trait to the reference since it exists, but this
+        # will allow access to the underlying item/object for any observes in the meantime.
+        for item in self.items:
+            if item['id'] == self.selected:
+                return item
+
+    @property
+    def selected_id(self):
+        return self.selected_item['id']
+
+    @property
+    def selected_obj(self):
+        return self.app.get_viewer_by_id(self.selected_id)
+
+    def _selected_changed(self, event):
+        if event['new'] not in self.ref_or_ids:
+            if self.selected in self.ids:
+                # provided id in place of ref
+                self.selected = self.ref_or_ids[self.ids.index(self.selected)]
+            else:
+                self._handle_default()
+                raise ValueError(f"{event['new']} not one of {self.ref_or_ids}")
+
+    def _handle_default(self):
+        if self.selected not in self.ref_or_ids:
+            # default to first entry, will trigger any observer on selected
+            self.selected = self.ref_or_ids[0] if len(self.items) else ""
+
+    def _on_viewers_changed(self):
```

Comment:
This goes with removal of `lambda` above.
```suggestion
    def _on_viewers_changed(self, *args):
```

Reply:
I dislike `*args`, but will change this to take `msg`.  The downside to doing this over the lambda is that internal calls will have to pass a message (or we can default to `None`, but then we need a check for `None` if we ever add code that does use the message object).